# Peer-graded Assignment: Build a Regression Model in Keras

## Concrete Strength Prediction - Part C

### Submitted by: Muhammad Jawad Bashir

**1. Importing Libararies & Dataset**

In [1]:
import pandas as pd
import numpy as np

In [2]:
concrete_data = pd.read_csv('https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DL0101EN/labs/data/concrete_data.csv')
concrete_data.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

In [3]:
concrete_data.isnull().sum()

Cement                0
Blast Furnace Slag    0
Fly Ash               0
Water                 0
Superplasticizer      0
Coarse Aggregate      0
Fine Aggregate        0
Age                   0
Strength              0
dtype: int64

**2. Separate into features & labels**

In [4]:
#Let's split the data into predictors and target
cd_columns = concrete_data.columns
predictors = concrete_data[cd_columns[cd_columns != 'Strength']]
target = concrete_data['Strength']
target.head()

0    79.99
1    61.89
2    40.27
3    41.05
4    44.30
Name: Strength, dtype: float64

**2a: Normalizing the Dataset***

In [5]:
pred_norm = (predictors - predictors.mean()) / predictors.std()
pred_norm.head()

Cement  Blast Furnace Slag   Fly Ash     Water  Superplasticizer  \
0  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
1  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
2  0.491187            0.795140 -0.846733  2.174405         -1.038638   
3  0.491187            0.795140 -0.846733  2.174405         -1.038638   
4 -0.790075            0.678079 -0.846733  0.488555         -1.038638   

   Coarse Aggregate  Fine Aggregate       Age  
0          0.862735       -1.217079 -0.279597  
1          1.055651       -1.217079 -0.279597  
2         -0.526262       -2.239829  3.551340  
3         -0.526262       -2.239829  5.055221  
4          0.070492        0.647569  4.976069

In [6]:
inp_cols = pred_norm.shape[1]
inp_cols

8

**3. Defining Keras Regression Model**

In [7]:
import keras
from keras.models import Sequential
from keras.layers import Dense

In [8]:
def model_reg():
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(inp_cols,)))
    model.add(Dense(1))
    model.compile(optimizer = 'adam', loss='mean_squared_error')
    return model

**4. Running the model for 50 times with given requirements in Assignment Part-C**

In [9]:
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error

In [10]:
mse_list = []
model = model_reg()
for i in range(0,50):
    X_train, X_test, y_train, y_test = train_test_split(pred_norm, target, test_size = 0.30, random_state=42)
    model.fit(X_train, y_train, epochs=100, verbose=0)
    y_pred = model.predict(X_test)
    mse = mean_squared_error(y_test,y_pred)
    mse_list.append(mse)

In [11]:
mse_list

[185.04506141301462,
 86.00656340102131,
 55.87256493374258,
 47.82517333947846,
 44.90074672385701,
 43.59956323614237,
 42.455772893697784,
 41.97453607244428,
 41.41123310867232,
 40.802247987774095,
 40.76650409882927,
 40.36372085519621,
 40.70540911794862,
 40.628077939026795,
 40.16610400754221,
 40.13195674801448,
 40.21870786057688,
 40.228790717999786,
 40.15507379183835,
 40.097205473300185,
 40.30099635907323,
 40.417283784787095,
 40.283826983303776,
 40.72175697329964,
 40.45164354256521,
 40.434161919021314,
 40.53933522083936,
 40.41883124308396,
 40.47192062761147,
 40.435279908496994,
 40.54774259290449,
 40.37709420574683,
 40.364302238557286,
 40.753236987793144,
 40.62075978146104,
 40.61583547560449,
 40.66961961478025,
 40.770828944487995,
 40.58951697988971,
 40.81911120485573,
 40.6579278810837,
 40.66779926726078,
 40.65653151176398,
 40.60599634321218,
 40.705072486935265,
 40.45004623513416,
 40.530799764986305,
 40.56443009370683,
 40.70840738909826,
 40.59

**5. MSE mean & Standard Deviation for Task-A**

In [12]:
print('Mean for Task-C: {}'.format(np.mean(mse_list)))
print('Standard Deviation for Task-C: {}'.format(np.std(mse_list)))

Mean for Task-C: 45.00204086494594
Standard Deviation for Task-C: 21.108337052990887


### Difference between MSEs of Part-B (nromalized dataset & 50 epochs) and Part-C(nromalized dataset + 100 epochs)

MSE in Part B = 50.019

MSE in Part C = 45.002

MSE further decreased by 5.017 after increasing the number of epochs to 100 in Part-C.